From 386c8660ae5bf7926d79bc2b4f5e6a5747a54156 Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Thu, 24 Oct 2024 16:54:37 -0700
Subject: [PATCH 23/28] PENDING: net: stmmac: Use new Qualcomm XPCS driver

Add Qualcomm XPCS(Physical Coding Sublayer) header
within EMAC glue driver and also make it call few
important callbacks functions exposed by Xpcs as
Emac driver is the centralized unit which is needed
to maintain proper sync between Ethernet, Serdes and
XPCS hardware blocks.

1) Create an Xpcs instance if "qcom-xpcs-handle" entry
is present within the ethernet dtsi node.
2) Call Xpcs link up as part of the mac link up call
flow and after the Serdes has been programmed.

Change-Id: I90dfa68bb13818ada8295b456fe04de7142bc4a5
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
---
 .../stmicro/stmmac/dwmac-qcom-ethqos.c        | 29 +++++++++++++++++++
 .../net/ethernet/stmicro/stmmac/stmmac_main.c |  3 ++
 include/linux/stmmac.h                        |  1 +
 3 files changed, 33 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
index bd04210c59fe1..7a6dfef023223 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
@@ -7,6 +7,7 @@
 #include <linux/platform_device.h>
 #include <linux/phy.h>
 #include <linux/phy/phy.h>
+#include <linux/pcs-xpcs-qcom.h>
 
 #include "stmmac.h"
 #include "stmmac_platform.h"
@@ -558,6 +559,8 @@ static void ethqos_fix_mac_speed_rgmii(void *bsp_priv,
 {
 	struct qcom_ethqos *ethqos = bsp_priv;
 	struct device *dev;
+	struct net_device *ndev = platform_get_drvdata(ethqos->pdev);
+	struct stmmac_priv *priv = netdev_priv(ndev);
 	unsigned int i;
 	u32 val;
 
@@ -623,6 +626,10 @@ static void ethqos_fix_mac_speed_rgmii(void *bsp_priv,
 		ethqos_dll_configure(ethqos);
 
 	ethqos_rgmii_macro_init(ethqos, speed);
+
+	if (priv->plat->qcom_pcs)
+		qcom_xpcs_link_up(priv->plat->qcom_pcs, mode, interface, speed,
+				  DUPLEX_FULL);
 }
 
 static void ethqos_pcs_set_inband(struct qcom_ethqos *ethqos, bool enable)
@@ -641,6 +648,8 @@ static void ethqos_fix_mac_speed_sgmii(void *bsp_priv,
 				       unsigned int mode)
 {
 	struct qcom_ethqos *ethqos = bsp_priv;
+	struct net_device *ndev = platform_get_drvdata(ethqos->pdev);
+	struct stmmac_priv *priv = netdev_priv(ndev);
 
 	switch (speed) {
 	case SPEED_2500:
@@ -659,6 +668,10 @@ static void ethqos_fix_mac_speed_sgmii(void *bsp_priv,
 	}
 
 	ethqos_pcs_set_inband(ethqos, interface == PHY_INTERFACE_MODE_SGMII);
+
+	if (priv->plat->qcom_pcs)
+		qcom_xpcs_link_up(priv->plat->qcom_pcs, mode, interface, speed,
+				  DUPLEX_FULL);
 }
 
 
@@ -667,6 +680,8 @@ static void ethqos_fix_mac_speed_usxgmii(void *bsp_priv,
 					 unsigned int mode)
 {
 	struct qcom_ethqos *ethqos = bsp_priv;
+	struct net_device *ndev = platform_get_drvdata(ethqos->pdev);
+	struct stmmac_priv *priv = netdev_priv(ndev);
 	unsigned int i;
 
 	qcom_ethqos_set_sgmii_loopback(ethqos, false);
@@ -748,6 +763,10 @@ static void ethqos_fix_mac_speed_usxgmii(void *bsp_priv,
 			"Invalid speed %d\n", speed);
 		return;
 	}
+
+	if (priv->plat->qcom_pcs)
+		qcom_xpcs_link_up(priv->plat->qcom_pcs, mode, interface, speed,
+				  DUPLEX_FULL);
 }
 
 static int qcom_ethqos_serdes_powerup(struct net_device *ndev, void *priv)
@@ -935,6 +954,7 @@ static int qcom_ethqos_hdma_cfg(struct platform_device *pdev, struct plat_stmmac
 static int qcom_ethqos_probe(struct platform_device *pdev)
 {
 	struct device_node *np = pdev->dev.of_node;
+	struct device_node *pcs_node;
 	const struct ethqos_emac_driver_data *data;
 	struct plat_stmmacenet_data *plat_dat;
 	struct stmmac_resources stmmac_res;
@@ -1066,6 +1086,15 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 		plat_dat->serdes_powerdown  = qcom_ethqos_serdes_powerdown;
 	}
 
+	if (of_property_present(dev->of_node, "qcom-xpcs-handle")) {
+		pcs_node = of_parse_phandle(dev->of_node, "qcom-xpcs-handle", 0);
+		plat_dat->qcom_pcs = qcom_xpcs_create(pcs_node, plat_dat->phy_interface);
+		if (IS_ERR_OR_NULL(plat_dat->qcom_pcs)) {
+			dev_warn(dev, "Qcom Xpcs not found\n");
+			return -ENODEV;
+		}
+	}
+
 	/* Enable TSO on queue0 and enable TBS on rest of the queues */
 	for (i = 1; i < plat_dat->tx_queues_to_use; i++)
 		plat_dat->tx_queues_cfg[i].tbs_en = 1;
diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
index 507e1c0023f7d..1fd259c46d905 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -1032,6 +1032,9 @@ static struct phylink_pcs *stmmac_mac_select_pcs(struct phylink_config *config,
 	    test_bit(interface, priv->integrated_pcs->pcs.supported_interfaces))
 		return &priv->integrated_pcs->pcs;
 
+	if (priv->plat->qcom_pcs)
+		return priv->plat->qcom_pcs;
+
 	return NULL;
 }
 
diff --git a/include/linux/stmmac.h b/include/linux/stmmac.h
index 599257ccbc6ce..e8ac260223b29 100644
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -326,6 +326,7 @@ struct plat_stmmacenet_data {
 	void (*get_queue_and_tc_from_vdma)(struct stmmac_priv *priv, u32 vdma_ch,
 					   unsigned long *queue_mask, u32 *tc);
 	int (*clks_config)(void *priv, bool enabled);
+	struct phylink_pcs *qcom_pcs;
 	int (*crosststamp)(ktime_t *device, struct system_counterval_t *system,
 			   void *ctx);
 	void (*dump_debug_regs)(void *priv);
